File: Angular/src/app/core/services/projects.service.ts
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';
import { ApiClientService } from './api-client.service';
import type {
  ChangeOrderSummaryDto,
  MapLocationDto,
  PagedResponse,
  ProjectDetailDto,
  ProjectKpisDto,
  ProjectMilestoneDto,
  ProjectSummaryDto,
  QueryParameters,
  RecentDocumentDto,
  RfiSummaryDto,
  RiskSummaryDto,
  SubmittalSummaryDto,
} from '../models/api.models';

@Injectable({ providedIn: 'root' })
export class ProjectsService {
  private api = inject(ApiClientService);

  getProjects(params?: QueryParameters): Observable<PagedResponse<ProjectSummaryDto>> {
    return this.api.getJson<PagedResponse<ProjectSummaryDto>>('projects', params);
  }

  getById(id: number): Observable<ProjectDetailDto> {
    return this.api.getJson<ProjectDetailDto>(`projects/${id}`);
  }

  getKpis(id: number): Observable<ProjectKpisDto> {
    return this.api.getJson<ProjectKpisDto>(`projects/${id}/kpis`);
  }

  getTopRisks(id: number, limit = 5): Observable<RiskSummaryDto[]> {
    return this.api.getJson<RiskSummaryDto[]>(`projects/${id}/top-open-risks`, { limit });
  }

  // NOTE: returns ProjectMilestoneDto[], NOT UpcomingMilestoneDto[] — a different shape
  // (plannedDate/status/owner vs dueDate/healthStatus) from the portfolio-wide report endpoint.
  getUpcomingMilestones(id: number, days = 60, limit = 10): Observable<ProjectMilestoneDto[]> {
    return this.api.getJson<ProjectMilestoneDto[]>(`projects/${id}/upcoming-milestones`, { days, limit });
  }

  getRecentDocuments(id: number, days = 90, limit = 10): Observable<RecentDocumentDto[]> {
    return this.api.getJson<RecentDocumentDto[]>(`projects/${id}/recent-documents`, { days, limit });
  }

  getRfis(id: number, limit = 50): Observable<RfiSummaryDto[]> {
    return this.api.getJson<RfiSummaryDto[]>(`projects/${id}/rfis`, { limit });
  }

  getSubmittals(id: number, limit = 50): Observable<SubmittalSummaryDto[]> {
    return this.api.getJson<SubmittalSummaryDto[]>(`projects/${id}/submittals`, { limit });
  }

  getChangeOrders(id: number, limit = 50): Observable<ChangeOrderSummaryDto[]> {
    return this.api.getJson<ChangeOrderSummaryDto[]>(`projects/${id}/change-orders`, { limit });
  }

  getLocations(): Observable<MapLocationDto[]> {
    return this.api.getJson<MapLocationDto[]>('projects/locations');
  }
}

File: Blazor/Core/Services/ProjectsService.cs
using Construction.Core.DTOs;

namespace Construction.Blazor.Core.Services;

public class ProjectsService(ApiClient api)
{
    public Task<PagedResponseDto<ProjectDto>> GetProjectsAsync(int page = 1, int pageSize = 50) =>
        api.GetJsonAsync<PagedResponseDto<ProjectDto>>("projects", new Dictionary<string, object?> { ["page"] = page, ["pageSize"] = pageSize });

    public Task<ProjectDto> GetByIdAsync(int id) =>
        api.GetJsonAsync<ProjectDto>($"projects/{id}");

    public Task<ProjectKpisDto> GetKpisAsync(int id) =>
        api.GetJsonAsync<ProjectKpisDto>($"projects/{id}/kpis");

    public Task<List<RiskDto>> GetTopRisksAsync(int id, int limit = 5) =>
        api.GetJsonAsync<List<RiskDto>>($"projects/{id}/top-open-risks", new Dictionary<string, object?> { ["limit"] = limit });

    // Distinct from ReportsService.GetUpcomingMilestonesAsync (portfolio-wide UpcomingMilestoneDto) —
    // this per-project endpoint returns MilestoneDto (own status/plannedDate/owner), not project health.
    public Task<List<MilestoneDto>> GetUpcomingMilestonesAsync(int id, int days = 60, int limit = 10) =>
        api.GetJsonAsync<List<MilestoneDto>>($"projects/{id}/upcoming-milestones", new Dictionary<string, object?> { ["days"] = days, ["limit"] = limit });

    public Task<List<RecentDocumentDto>> GetRecentDocumentsAsync(int id, int days = 90, int limit = 10) =>
        api.GetJsonAsync<List<RecentDocumentDto>>($"projects/{id}/recent-documents", new Dictionary<string, object?> { ["days"] = days, ["limit"] = limit });

    public Task<List<RfiSummaryDto>> GetRfisAsync(int id, int limit = 50) =>
        api.GetJsonAsync<List<RfiSummaryDto>>($"projects/{id}/rfis", new Dictionary<string, object?> { ["limit"] = limit });

    public Task<List<SubmittalSummaryDto>> GetSubmittalsAsync(int id, int limit = 50) =>
        api.GetJsonAsync<List<SubmittalSummaryDto>>($"projects/{id}/submittals", new Dictionary<string, object?> { ["limit"] = limit });

    public Task<List<ChangeOrderSummaryDto>> GetChangeOrdersAsync(int id, int limit = 50) =>
        api.GetJsonAsync<List<ChangeOrderSummaryDto>>($"projects/{id}/change-orders", new Dictionary<string, object?> { ["limit"] = limit });

    public Task<List<MapLocationDto>> GetLocationsAsync() =>
        api.GetJsonAsync<List<MapLocationDto>>("projects/locations");
}
